### PR TITLE
84 implement posts api for course forums

### LINKED ISSUE #84: Implement Posts API for Course Forums
## Overview
Create CRUD endpoints for managing course discussion posts with user tracking, edit history, and threading

## Database Tables
- [x] `posts` table:
  ```
  - id: UUID
  - course_id: UUID
  - title: string
  - content: text
  - parent_id: UUID (optional)
  - created_by: UUID
  - status: enum (active, deleted)
  - timestamps
  ```

- [x] `post_edits` table:
  ```
  - id: UUID
  - post_id: UUID
  - edited_by: UUID
  - previous_content: text
  - new_content: text
  - edit_reason: text (optional)
  - timestamps
  ```

## API Endpoints
Under `/courses/{course_id}/posts`:

- [x] `POST`: create new post
  ```json
  {
    "title": "string",
    "content": "string",
    "parent_id": "UUID (optional)",
    "tags": "UUID[]"
  }
  ```
  - Track creator
  - Create embedding job

- [x] `GET`: list posts with metadata
  - Include post creators
  - Show last editor if edited
  - Filter/sort options
  - Edit history access

- [x] `PATCH /{post_id}`: update post
  - Save edit history
  - Track editor
  - Require edit reason for others' posts
  - Permission checks

- [x] `DELETE /{post_id}`: Hard delete (for now)
  - ~~Track who deleted~~
  - ~~Keep history~~

Tasks:
- [x] Write migrations
- [x] Add user tracking
- [x] Implement edit history
- [x] Add permission logic

### PR DESCRIPTION
**Pull request by bcw117**
## Proposed Changes

- closes #84

### CODE REVIEW

File: backend/routers/routes/posts.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+class PostEdit(BaseModel):
+    new_content: str
+    edit_reason: str | None = None
+
+
+@post_router.post("/courses/{course_id}/posts")
+async def create_post(course_id: str, post_info: Post):
+    # Test string for user_id
+    user_id = "686fa92e-696d-4bf8-9b27-1dbc626fc5e9"
```
Comment (armintalaie) [inline backend/routers/routes/posts.py]: for the middleware - we can access the user now - check `auth.py`

File: backend/main.py [hunk 2]
```diff
@@ -4,11 +4,13 @@
 from fastapi import FastAPI
 from middleware.auth import AuthMiddleware
 from routers.routes.courses import course_router
+from routers.routes.posts import post_router
 from routers.routes.users import user_router
 
 app = FastAPI()
 app.include_router(user_router)
 app.include_router(course_router)
+app.include_router(post_router)
```
Comment (armintalaie) [inline backend/main.py]: - Pull from `dev` and see how the `courses.py` is setting `documents` as its subpath

File: backend/routers/routes/posts.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    title: str
+    content: str
+    parent_id: str | None = None
+    tags: list[str]
+
+
+class PostEdit(BaseModel):
+    new_content: str
+    edit_reason: str | None = None
+
```
Comment (armintalaie) [inline backend/routers/routes/posts.py]: thoughts on moving this to the `models` folder - (i noticed we were adding a bunch of these classes)
https://github.com/ubclaunchpad/forum/tree/dev/backend/models

**Review by armintalaie — CHANGES_REQUESTED** [review]
Awesome work @bcw117 - some tiny tweaks :)

File: backend/models/post_models.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from enum import Enum
+
+from pydantic import BaseModel
+
+
+class Post(BaseModel):
+    """Model for post metadata"""
+
+    title: str
```
Comment (armintalaie) [inline backend/models/post_models.py:11]: can we add field validator here? https://docs.pydantic.dev/latest/concepts/validators/#field-validators
i.e. "has to be at least 5 characters?"

File: backend/models/post_models.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    content: str
+    parent_id: str | None = None
+    tags: list[str]
```
Comment (armintalaie) [inline backend/models/post_models.py:14]: validator for this. all tags should be alphanumeric or have slash. Use gpt or something to make a regex for it. Basically 
`<word>/<word>` or `<word>` 

Bad:
- ""
- "assignments//1"
- "assignment??/"
- "_a"
- ..

File: backend/routers/routes/posts.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+from typing import Annotated, Literal
+
+from crud import post_crud, user_crud
+from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response
+from models.post_models import FilterParams, Post, PostEdit
+
+post_router = APIRouter()
+
+
+@post_router.post("")
```
Comment (armintalaie) [inline backend/routers/routes/posts.py]: for the endpoints make a response model:
`@document_router.post("", response_model=DocumentResponse)`

File: backend/database/migrations/1731975568326.sql [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+drop table if exists public.posts;
+drop table if exists public.post_edits;
+create table public.posts (
+  id uuid default uuid_generate_v4(),
+  course_id uuid,
+  title text,
+  content text,
+  parent_id uuid,
+  created_by uuid,
```
Comment (armintalaie) [inline backend/database/migrations/1731975568326.sql]: this should reference a user (add a foreign constraint) (on delete do nothing)

**Review by armintalaie — APPROVED** [review]
Good work. Thanks for adding proper status codes. 

- can u add response model for the endpoints (helps for FE folks and also throws an error on backend in response is not how it should be structured)
- add foreign ref to the posts table to connect to the user

File: backend/database/migrations/1731975568326.sql (hunk 7 above)
Comment (bcw117) [reply backend/database/migrations/1731975568326.sql]: Also added foreign constraint _edit_by_ in post_edits

**Pull request merged**